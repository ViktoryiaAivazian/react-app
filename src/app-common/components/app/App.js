import React, { Component } from "react";
import {BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import { connect } from 'react-redux';
import HomePage from '../pages/home-page/HomePage.js';
import MoviePage from '../pages/movie-page/MoviePage.js';
import ErrorBoundary from '../errorBoundary/ErrorBoundary.js';
import { moviesFetchData, getMovieAction, sortMoviesAction } from './app.action.js';
import "./App.pcss";

class App extends Component {

    constructor(props){
        super(props);

        this.onChangeSort = this.onChangeSort.bind(this);
        this.getCurrentMovie = this.getCurrentMovie.bind(this);
    }

    onChangeSort(e){
        this.props.setTypeSort(e.target.value, this.sortMovies(e.target.value));
    }

    sortMovies(sortType){
        let sortMoviesList = this.props.movies;

        switch (sortType) {
            case 'rating':
                sortMoviesList = sortMoviesList.sort((a, b) => (b.id - a.id));
                break;
            case 'date':
                sortMoviesList = sortMoviesList.sort((a, b) => (new Date(b.release_date) - new Date(a.release_date)));
                break;
        }

        return sortMoviesList
    }

    getCurrentMovie(id) {
        this.props.getMovieDetails(id, this.filterMoviesByGenres(id));
    }

    filterMoviesByGenres(id){

        let movies = this.props.copyMovies;
        let currentGenre = movies.filter(item => item.id === id).map((key) => key.genres);

        let sort = movies.filter((film) => {
            let genres = film.genres.join();
            return currentGenre.reduce((res, genre) => {
                return !!genres.includes(genre) || res;
            }, false);
        });

        return {
            sort,
            currentGenre
        };
    }

    componentDidMount(){
        this.props.fetchData('http://reactjs-cdp.herokuapp.com/movies');
    }

    render() {

        const {movies, currentMovie, currentMoviesGenres, selectedMovieGenre, copyMovies, countMovies, sortBy} = this.props;

        return (
            <ErrorBoundary>
                <Router>
                    <Switch>
                        <Route exact path="/">
                            <HomePage
                                movies={movies}
                                copyMovies={copyMovies}
                                countMovies={countMovies}
                                sortBy={sortBy}
                                getCurrentMovie={this.getCurrentMovie}
                                onChangeSort={this.onChangeSort}
                            />
                        </Route>
                        <Route exact path="/movie-page/:id">
                            <MoviePage
                                movies={movies}
                                currentMoviesGenres={currentMoviesGenres}
                                copyMovies={copyMovies}
                                selectedMovieGenre={selectedMovieGenre}
                                getCurrentMovie={this.getCurrentMovie}
                                currentMovie={currentMovie}/>
                        </Route>
                    </Switch>
                </Router>
            </ErrorBoundary>
        );
    }
}

// mapDispatchToProps , mapStateToProps связывают компоненты со стором
// mapStateToProps передает стор в пропсы
// mapStateToProps должна возвращать объект
const mapStateToProps = (state) => {
   return {
       movies: state.app.movies,
       copyMovies: state.app.copyMovies,
       sortBy: state.app.sortBy,
       currentMovie: state.app.currentMovie,
       currentMoviesGenres: state.app.currentMoviesGenres,
       selectedMovieGenre: state.app.selectedMovieGenre,
   }
};

const mapDispatchToProps = (dispatch) => {
    return {
        fetchData: url => dispatch(moviesFetchData(url)),
        getMovieDetails: (currentMovie, obj) => {
            dispatch(getMovieAction(currentMovie, obj))
        },
        setTypeSort: (sortBy, movies) => {
            dispatch(sortMoviesAction(sortBy, movies))
        },
    }
};

// connect используется для связывания хранилища Redux с компонентами React.
export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(App);


