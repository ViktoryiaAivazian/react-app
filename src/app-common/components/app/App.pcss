#root {
    height: 100%;
}
* {
    box-sizing: border-box;
}
html,
body {
    height: 100%;
}
body {
    font-family: 'Noto Sans JP', sans-serif;
    color: #221f1f;
    font-size: 16px;
}
a {
    text-decoration: none;
}
.container {
    margin: 0 auto;
    max-width: 1100px;
    padding: 0 15px;
}
.main {
    height: 100%;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
}
.content {
    flex-grow: 1;
    &__outer {
        display: flex;
        flex-direction: row;
        justify-content: flex-start;
        margin: 0 -30px;
        flex-wrap: wrap;
    }
    &__itm {
        padding: 0 30px;
        width: 33%;
        margin-bottom: 30px;
    }
    &__top {
        background: #f3f3f3;
        padding: 20px 0;
        margin-bottom: 40px;
    }
    &__bottom {
        .container {
            max-width: 880px;
        }
    }
}