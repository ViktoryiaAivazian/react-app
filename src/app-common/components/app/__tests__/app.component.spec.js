import React from "react";
import { shallow } from "enzyme";
import configureMockStore from "redux-mock-store"; // для написания системных тестов у контейнеров, где мы проверяем функционал в связке.
import { Provider } from "react-redux";
import toJson from 'enzyme-to-json';
import App from "../App.js";

const mockStore = configureMockStore();
const store = mockStore({});

const state = {
    movies: [],
    copyMovies: []
};

describe('App component', () => {

    describe('with HomePage component', () => {
        it('should render <App> with <HomePage>', () => {
            const component = shallow(
                <Provider store={store}>
                    <App {...state}/>
                </Provider>);

            expect(toJson(component)).toMatchSnapshot();
        });
    });

    describe('with MoviePage component', () => {
        it('should render <App> with < MoviePage >', () => {
            const component = shallow(
                <Provider store={store}>
                    <App {...state}/>
                </Provider>);

            component.update();
            expect(toJson(component)).toMatchSnapshot();
        });
    });
});




