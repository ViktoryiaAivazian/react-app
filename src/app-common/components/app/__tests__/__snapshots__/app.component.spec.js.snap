// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component with HomePage component should render <App> with <HomePage> 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(App)
    copyMovies={Array []}
    movies={Array []}
  />
</ContextProvider>
`;

exports[`App component with MoviePage component should render <App> with < MoviePage > 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(App)
    copyMovies={Array []}
    movies={Array []}
  />
</ContextProvider>
`;
